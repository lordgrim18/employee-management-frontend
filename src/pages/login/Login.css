.login-page {
    display: flex;
    flex-direction: row;
    height: 100vh;
}

.side-banner {
    /* flex: 1; */
    width: 50%;
    background-color: #dbf3fc;
    display: flex;
    align-items: center;
    justify-content: center;
    margin: 20px;
}

.side-banner img {
    border-radius: 50%;
    width: 80%;
    aspect-ratio: 1/1;
    object-fit: cover;
}

.login-form {
    /* flex: 1; */
    width: 50%;
    display: flex;
    justify-content: space-around;
    align-items: center;
}

.login-form-content {
    width: max-content;
    display: flex;
    flex-direction: column;
    gap: 30px;
}

.login-form img {
    height: 55px;
    width: 60%;
}

.login-form__form {
    display: flex;
    flex-direction: column;
    gap: 10px
}

.form-element--login {
    position: relative;
    width: 45vb;
}

.form-element--login input {
    height: 45px;
}

.form-element--login label {
    position: absolute;
    left: 15px;
    top: 10px;
}

.form-element--login input:focus {
    color: #03AEEE;
    border: 2px solid rgb(211, 208, 208);
}

.form-element--login input:focus+.form-element label {
    top: -10px;
    color: #03AEEE;
    font-size: 12px;
}

.button--login {
    height: 50px;
    color: white;
    background-color: #03AEEE;
    box-shadow: 1px 1px 5px #03AEEE;
    border-color: transparent;
}

.login--clear-button {
    padding: 10px;
}