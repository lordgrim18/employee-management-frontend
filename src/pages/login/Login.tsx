import "./Login.css";
import LoginInput from "../../components/LoginInput/LoginInput";
import Button from "../../components/Button/Button";

import loginSideLogo from "../../assets/kv-login.jpeg";
import KVlogo from "../../assets/kv-logo.png";
import { useEffect, useRef, useState, type SetStateAction } from "react";

import useMousePosition from "../../hooks/useMousePosition";
import { Navigate, useNavigate } from "react-router-dom";
import { useLoginMutation } from "../../api-service/auth/login.api";

const sampleUser = {
  username: "John",
  password: "Password"
}

const Login = () => {
  const [login] = useLoginMutation();
  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");
  const [showPassword, setShowPassword] = useState(localStorage.getItem("showPassword") === "true");
  const [error, setError] = useState({
    error: "",
    usernameError: "",
    passwordError: ""
  })

  const navigate = useNavigate()

  const usernameRef = useRef<HTMLInputElement>(null);
  const mousePosition = useMousePosition();

  const updateUsername = (event: {
    target: { value: SetStateAction<string> };
  }) => {
    setUsername(event.target.value);
  };


  useEffect(() => {
    if (username.length < 4 && username.length > 0) {
      setError((prevError) => ({
        ...prevError,
        usernameError: "Username should be atleast 4 characters long"
      }));
      return;
    }
    setError((prevError) => ({
      ...prevError,
      usernameError: ""
    }));
    return;
  }, [username]);

  useEffect(() => {
    if (password.length < 5 && password.length > 0) {
      setError((prevError) => ({
        ...prevError,
        passwordError: "Password should be atleast 5 characters long"
      }));
      return;
    }
    setError((prevError) => ({
      ...prevError,
      passwordError: ""
    }));
    return;
  }, [password]);

  useEffect(() => {
    setError((prevError) => ({
        ...prevError,
        error: ""
      }))
  }, [username, password])

  useEffect(() => {
    if (usernameRef.current) {
      usernameRef.current.focus();
    }
  }, []);

  // const loginUser = () => {
  //   if (username === sampleUser.username && password === sampleUser.password) {
  //     localStorage.setItem("isLoggedIn", "true");
  //     setError((prevError) => ({
  //       ...prevError,
  //       error: ""
  //     }))
  //     navigate("/employees")
  //   } else {
  //     setError((prevError) => ({
  //       ...prevError,
  //       error: "Incorrect username or password"
  //     }))
  //   }
  // }

  const loginUser = async () => {
    const response = await login({ email: username, password: password });
    localStorage.setItem("token", response.data ? response.data.accessToken : '');
    navigate('/employees')
  }

  const isLoggedIn = () => {
    const token = localStorage.getItem("token");
    return token ? true : false 
  };

  if (isLoggedIn()) {
    return <Navigate to='/employees' />
  }

  return (
    <div className="login-page">
      <div className="side-banner">
        <img src={loginSideLogo} alt="side-banner" />
      </div>
      <div className="login-form">
        <form className="login-form-content">
          <div>
            <h3>Mouse x - {mousePosition.x}</h3>
            <h3>Mouse y - {mousePosition.y}</h3>
          </div>
          <img src={KVlogo} alt="kv-logo" className="kv-logo" />
          <div className="login-form__form">
            <LoginInput
              id="login-username"
              label="Username"
              value={username}
              onChange={updateUsername}
              ref={usernameRef}
              endAdornment={
                <button
                  className="login--clear-button"
                  type="button"
                  disabled={username.length === 0}
                  onClick={() => setUsername("")}
                >
                  Clear
                </button>
              }
            />
            {error.usernameError && <span style={{ color: 'red', fontSize: '12px' }}>{error.usernameError}</span>}

            <LoginInput
              id="login-password"
              label="Password"
              type={showPassword ? "text" : "password"}
              value={password}
              onChange={(event) => setPassword(event.target.value)}
              endAdornment={
                <button
                  className="login--clear-button"
                  type="button"
                  disabled={password.length === 0 }
                  onClick={() => setPassword("")}
                >
                  Clear
                </button>
              }
            />
            {error.passwordError && <span style={{ color: 'red', fontSize: '12px' }}>{error.passwordError}</span>}

            <LoginInput
              id="login-show-password"
              label="Show Password"
              type="checkbox"
              // {setShowPassword ? "checked" : }
              onChange={() => setShowPassword((prev) => {
                localStorage.setItem("showPassword", String(!prev))
                return !prev
              })}
              checked={showPassword}
              commonClass="form-element-checkbox"
              required={false}
            />
          </div>
          <Button type="button" onClick={loginUser} buttonName="Logging in" variant="login" />
          {error.error && <span style={{ color: 'red', fontSize: '12px' }}>{error.error}</span>}
        </form>
      </div>
    </div>
  );
};

export default Login;
