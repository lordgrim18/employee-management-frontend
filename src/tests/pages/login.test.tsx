import { describe, it, expect, beforeEach, vi } from "vitest";
import { render, screen, waitFor } from "@testing-library/react";
import { Provider } from "react-redux";
import { configureStore } from "@reduxjs/toolkit";
import userEvent from "@testing-library/user-event";
import Login from "../../pages/login/Login"; // Mock react-router-dom's useNavigate

const mockNavigate = vi.fn();
vi.mock("react-router-dom", async () => {
  const actual = await vi.importActual("react-router-dom");
  return {
    ...actual,
    useNavigate: () => mockNavigate,
  };
});

// Mock the store
const mockStore = configureStore({
  reducer: {
    employee: (state = { employees: [] }) => state,
    department: (state = { departments: [] }) => state,
    employeeApi: (state = {}) => state,
  },
});

// Mock the login mutation
const mockLoginMutation = vi.fn();
vi.mock("../../api-service/auth/login.api", () => ({
  useLoginMutation: () => [
    () => ({
      unwrap: () => mockLoginMutation(),
    }),
    { isLoading: false },
  ],
}));

describe("Login Page", () => {
  const renderLogin = () => {
    return render(
      <Provider store={mockStore}>
        <Login />
      </Provider>
    );
  };
  beforeEach(() => {
    vi.clearAllMocks();
  });
  it("should match snapshot", () => {
    const { container } = renderLogin();
    expect(container).toMatchSnapshot();
  });

  it("should render login form with all required elements", () => {
    renderLogin();

    // Check if username input is present
    const usernameInput = screen.getByLabelText("Username");
    expect(usernameInput).toBeInTheDocument();

    // Check if password input is present
    const passwordInput = screen.getByLabelText("Password");
    expect(passwordInput).toBeInTheDocument();

    // Check if submit button is present
    const submitButton = screen.getByRole("button", {name: /Log in/i });
    // const submitButton = screen.getByText(/Logging in/i)
    expect(submitButton).toBeInTheDocument();
    expect(submitButton).toHaveAttribute("type", "button");
  });

  it("should show error when username is too small", async () => {
    renderLogin();
    const usernameInput = screen.getByLabelText("Username");    // Type a username longer than 30 characters
    await userEvent.type(usernameInput, "a".repeat(3));    // Check if error message is displayed
    expect(
      screen.getByText("Username should be atleast 4 characters long")
    ).toBeInTheDocument();
  });

  it("should handle successful login", async () => {
    mockLoginMutation.mockResolvedValueOnce({ token: "fake-token" });

    renderLogin();

    // Fill in the form
    const usernameInput = screen.getByLabelText("Username");
    const passwordInput = screen.getByLabelText("Password");
    const submitButton = screen.getByRole("button", { name: /Log in/i });

    await userEvent.type(usernameInput, "test@example.com");
    await userEvent.type(passwordInput, "password123");

    // Submit the form
    await userEvent.click(submitButton);

    // Wait for the login mutation to resolve and navigation to occur
    await waitFor(() => {
      expect(mockLoginMutation).toHaveBeenCalled();
      expect(mockNavigate).toHaveBeenCalledWith("/employees");
    });
  });

});
