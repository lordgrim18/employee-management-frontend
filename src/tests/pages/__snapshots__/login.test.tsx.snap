// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Login Page > should match snapshot 1`] = `
<div>
  <div
    class="login-page"
  >
    <div
      class="side-banner"
    >
      <img
        alt="side-banner"
        src="/src/assets/kv-login.jpeg"
      />
    </div>
    <div
      class="login-form"
    >
      <form
        class="login-form-content"
      >
        <div>
          <h3>
            Mouse x - 
            0
          </h3>
          <h3>
            Mouse y - 
            0
          </h3>
        </div>
        <img
          alt="kv-logo"
          class="kv-logo"
          src="/src/assets/kv-logo.png"
        />
        <div
          class="login-form__form"
        >
          <div
            class="form-element-parent--login-username"
            style="display: flex; flex-direction: row; gap: 10px;"
          >
            <div
              class="form-element form-element--login form-element--login-username"
            >
              <input
                id="login-username"
                placeholder=""
                required=""
                type="text"
                value=""
              />
              <label
                for="login-username"
              >
                Username
              </label>
            </div>
            <button
              class="login--clear-button"
              disabled=""
              type="button"
            >
              Clear
            </button>
          </div>
          <div
            class="form-element-parent--login-password"
            style="display: flex; flex-direction: row; gap: 10px;"
          >
            <div
              class="form-element form-element--login form-element--login-password"
            >
              <input
                id="login-password"
                placeholder=""
                required=""
                type="password"
                value=""
              />
              <label
                for="login-password"
              >
                Password
              </label>
            </div>
            <button
              class="login--clear-button"
              disabled=""
              type="button"
            >
              Clear
            </button>
          </div>
          <div
            class="form-element-parent--login-show-password"
            style="display: flex; flex-direction: row; gap: 10px;"
          >
            <div
              class="form-element-checkbox form-element--login form-element--login-show-password"
            >
              <input
                id="login-show-password"
                placeholder=""
                type="checkbox"
              />
              <label
                for="login-show-password"
              >
                Show Password
              </label>
            </div>
          </div>
        </div>
        <button
          class="custom-button button--login"
          disabled=""
          type="button"
        >
          Log in
        </button>
      </form>
    </div>
  </div>
</div>
`;
